addon-test-support/@ember/test-helpers/-internal/helper-hooks.ts

```diff
@@ -0,0 +1,62 @@
+type Hook = (...args: any[]) => void | Promise<void>;
+type HookUnregister = {
+  unregister: () => void;
+};
+
+const registeredHooks = new Map<string, Set<Hook>>();
+
+/**
+ *
+ * @param {string} helperName The name of the test helper in which to run the hook.
+ * @param {string} order The order, either `start` or `end`, to run the hook.
+ * @returns {string} The compound key for the helper.
+ */
+function getHelperKey(helperName: string, order: string) {
+  return `${helperName}:${order}`;
+}
+
+/**
+ * Registers a hook function to be run during the invocation of a test helper.
+ *
+ * @public
+ * @param {string} helperName The name of the test helper in which to run the hook.
+ * @param {string} order The order, either `start` or `end`, to run the hook.
+ * @param {Function} hook The hook function to run when the test helper is invoked.
+ * @returns {HookUnregister} An object containing an unregister function that will unregister
+ *                           the specific hook registered to the helper.
+ */
+export function registerHook(helperName: string, order: string, hook: Hook): HookUnregister {
```

Comment:
I like the second arg, but I'm worried about the name here. Can we rename it from `order` to `label`? 

Specifically, I'd like to add something that isn't "before" / "after" related to `waitUntil` (e.g. `runHooks('waitUntil', 'check')` each time we iterate), so naming this "order" is a bit bad.

Reply:
I can do that. Label seems fine, though a tad more vague, but it is a bit more flexible in its usage.